### PR TITLE
Features/segments

### PR DESCRIPTION
**Pull request by jeferson-cleancloud**
Included API for segments in order to aggregate multiple context contraints to enable reusage of constraints per business among projects.

@vctrodrigues  to setup locally

```
# Install Postgres
docker pull postgres
docker run --name postgres -e POSTGRES_PASSWORD=postgres -p 5432:5432 -d postgres

docker exec -it postgres psql -U postgres
psql -h localhost -p 5432 -U postgres
CREATE DATABASE unleash;
CREATE USER unleash_user WITH PASSWORD 'password';
ALTER DATABASE unleash OWNER TO unleash_user;

# Initiate project
npm install
npm run dev

# Admin Login
# user: admin
# password: unleash4all
```

Tests with segments

List on project and menus
Segments-0-

Create segment on project
Segments-create-

Edit segment
segment-edit

Segments outside project
segments-menu

Segments on different projects
segments-different-projects

Segments delete
Segment-delete

### CODE REVIEW

File: src/lib/routes/admin-api/segments.ts [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+import { Logger } from '../../logger';
+import { OpenApiService } from '../../services/openapi-service';
+import { IAuthRequest } from '../unleash-types';
+import { createRequestSchema } from '../../openapi/util/create-request-schema';
+import { createResponseSchema } from '../../openapi/util/create-response-schema';
+import {
+    segmentSchema,
+    SegmentSchema,
+} from '../../openapi/spec/segment-schema';
+// import {
```
Comment (jeferson-cleancloud) [inline src/lib/routes/admin-api/segments.ts]: @vctrodrigues uncomment this block for testing

File: src/lib/routes/admin-api/segments.ts [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+                        200: emptyResponse,
+                    },
+                }),
+            ],
+        });
+    }
+
+    async getSegments(
+        req: IAuthRequest,
+        res: Response<SegmentSchema[]>,
```
Comment (jeferson-cleancloud) [inline src/lib/routes/admin-api/segments.ts]: @vctrodrigues update SegmentSchema[] to SegmentsSchema

File: src/lib/routes/admin-api/segments.ts [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+    ): Promise<void> {
+        const allSegments = await this.service.getAll();
+
+        this.openApiService.respondWithValidation(
+            200,
+            res,
+            segmentSchema.$id,
+            serializeDates(allSegments),
```
Comment (jeferson-cleancloud) [inline src/lib/routes/admin-api/segments.ts]: @vctrodrigues update to {version: 1, segments: serializeDates(allSegments) },

File: src/lib/routes/admin-api/segments.ts [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+            200,
+            res,
+            segmentSchema.$id,
+            serializeDates(segment),
+        );
+    }
+
+    async getSegmentStrategies(
+        req: IAuthRequest<ISegmentParam, unknown, unknown, IArchivedQuery>,
+        res: Response<FeatureStrategyBySegmentSchema>,
```
Comment (jeferson-cleancloud) [inline src/lib/routes/admin-api/segments.ts:210]: @vctrodrigues here the same problem array vs object

**Review by vctrodrigues — APPROVED** [review]

**Pull request merged**